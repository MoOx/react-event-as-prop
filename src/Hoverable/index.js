// @flow

import React, { Component } from "react"

type PropsType = {
  onMouseEnter?: Function | boolean,
  onMouseLeave?: Function | boolean,
}

type StateType = {
  hovered: boolean,
}

// @todo try a more complex type for ComposedComponent + return type
export default (ComposedComponent: Function) => {
  class Hoverable extends Component<void, PropsType, StateType> {
    props: PropsType

    state: StateType = {
      hovered: false,
    }

    handleMouseEnter: Function = (event: SyntheticEvent): void => {
      this.setState({ hovered: true })
      if (typeof this.props.onMouseEnter === "function") {
        this.props.onMouseEnter(event)
      }
    }

    handleMouseLeave: Function = (event: SyntheticEvent): void => {
      this.setState({ hovered: false })
      if (typeof this.props.onMouseLeave === "function") {
        this.props.onMouseLeave(event)
      }
    }

    render(): React$Element<any> {
      return (
        <ComposedComponent
          { ...this.props }
          { ...this.state }
          onMouseEnter={ this.handleMouseEnter }
          onMouseLeave={ this.handleMouseLeave }
        />
      )
    }
  }

  return Hoverable
}
