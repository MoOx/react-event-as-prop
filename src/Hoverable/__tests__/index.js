import React from "react"
import renderer from "react-test-renderer"

import Hoverable from ".."

const Compo = ({ hovered, ...props }: { hovered: boolean }) => (
  <span { ...props }>{ hovered ? "hovered" : "nope" }</span>
)
const Composed = Hoverable(Compo)

test("hovered props changes when hovered", () => {
  const component = renderer.create(
    <Composed
      test
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  tree.props.onMouseEnter()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  tree.props.onMouseLeave()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

let flag = 0
const handleMouseEnter = () => flag = 1
const handleMouseLeave = () => flag = 2

test("hovered props changes when hovered and execute on{Event}", () => {
  const component = renderer.create(
    <Composed
      test
      onMouseEnter={ handleMouseEnter }
      onMouseLeave={ handleMouseLeave }
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(flag).toEqual(0)

  tree.props.onMouseEnter()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(flag).toEqual(1)

  tree.props.onMouseLeave()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(flag).toEqual(2)
})
