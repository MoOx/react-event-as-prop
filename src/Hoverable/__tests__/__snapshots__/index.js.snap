exports[`test hovered props changes when hovered 1`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  test={true}>
  nope
</span>
`;

exports[`test hovered props changes when hovered 2`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  test={true}>
  hovered
</span>
`;

exports[`test hovered props changes when hovered 3`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  test={true}>
  nope
</span>
`;

exports[`test hovered props changes when hovered and execute on{Event} 1`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  test={true}>
  nope
</span>
`;

exports[`test hovered props changes when hovered and execute on{Event} 2`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  test={true}>
  hovered
</span>
`;

exports[`test hovered props changes when hovered and execute on{Event} 3`] = `
<span
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  test={true}>
  nope
</span>
`;
